haystack/telemetry/_telemetry.py

```diff
@@ -144,18 +145,17 @@ def pipeline_running(pipeline: "Pipeline") -> Optional[Tuple[str, Dict[str, Any]
     pipeline._last_telemetry_sent = datetime.datetime.now()
 
     # Collect info about components
-    pipeline_description = pipeline.to_dict()
     components: Dict[str, List[Dict[str, Any]]] = defaultdict(list)
-    for component_name, component in pipeline_description["components"].items():
-        instance = pipeline.get_component(component_name)
+    for component_name, instance in pipeline.walk():
+        component_qualified_class_name = generate_qualified_class_name(type(instance))
         if hasattr(instance, "_get_telemetry_data"):
             telemetry_data = getattr(instance, "_get_telemetry_data")()
             try:
-                components[component["type"]].append({"name": component_name, **telemetry_data})
+                components[component_qualified_class_name].append({"name": component_name, **telemetry_data})
             except TypeError:
-                components[component["type"]].append({"name": component_name})
+                components[component_qualified_class_name].append({"name": component_name})
```

Comment:
Why do we need exception handling here? Looks like we expect the method to return a dict, so let's just check the type of the return value, raise an error if it's not a Dict and then unpack it.

Reply:
We have the guiding principle that telemetry should never let the execution of the user's pipeline fail. That's why we use a wrapper that catches any error raised by telemetry and sends a debug log message instead: https://github.com/deepset-ai/haystack/blob/312ba9ebf038cd3d04aaa2bf77e00dc1c8c10e9a/haystack/telemetry/_telemetry.py#L116

I changed the code here so that it raises a TypeError now. This error will be caught and become a log message then and no telemetry event will be sent. With the previous version of the code we sent an event but just without the telemetry_data.